highfidelity/hifi
# Material Entities can specify face culling method, support GLTF doubleSided property

(contains changes from https://github.com/highfidelity/hifi/pull/16528, merge that first)

labels: blocked by another PR, open source, engine, pre merge testing required, development ready

https://roadmap.highfidelity.com/open-source-project-proposals/p/let-material-entities-specify-backface-culling-method

* hifi_pbr material entities support an enum, cullFaceMode ("CULL_NONE," "CULL_FRONT," or "CULL_BACK").  works on models (entities/avatars) and shape entities
* gltf models support the [doubleSided](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#double-sided) property

Test plan:
- Full nitpick run should pass.
- Run [this](https://gist.githubusercontent.com/HifiExperiments/8b7e181d8d99e07b3211048154b3de14/raw/a667bdfd5a3cea03088bbd88e8cc181d425ca807/CullFaceModeTest.js).  You'll see:
![image](https://user-images.githubusercontent.com/53453710/69679385-d7190c00-105c-11ea-9c1c-49c8bc045f09.png)
- Press "u", you'll see this:
![image](https://user-images.githubusercontent.com/53453710/69679414-e4ce9180-105c-11ea-99ca-b744ced08ccb.png)
- Press "u", you'll see this (if you move your camera to be inside the sphere, you'll see the inside of it):
![image](https://user-images.githubusercontent.com/53453710/69679476-0e87b880-105d-11ea-9285-75b020f02cb8.png)
- Press "u", you'll see this (if you move your camera to be inside the sphere, you won't see the inside of it):
![image](https://user-images.githubusercontent.com/53453710/69679476-0e87b880-105d-11ea-9285-75b020f02cb8.png)
- Press "u", you'll see this:
![image](https://user-images.githubusercontent.com/53453710/69679537-35de8580-105d-11ea-9fff-0202d89ce8a7.png)
- Press "u", you'll see this (if you move your camera to be inside the gnome, you'll see the inside of it):
![image](https://user-images.githubusercontent.com/53453710/69679547-3f67ed80-105d-11ea-867c-ffd485a5c8b8.png)
- Press "u", you'll see this (if you move your camera to be inside the gnome, you won't see the inside of it):
![image](https://user-images.githubusercontent.com/53453710/69679547-3f67ed80-105d-11ea-867c-ffd485a5c8b8.png)
